File: max_product_3_numbers.c
// 628. Maximum Product of Three Numbers
#include "leetcode.h"

/*
 * given an integer array 'nums', find three numbers whose product is maximum
 * and return the maximum product
 */

int maximumProduct(int *nums, int numsSize) {
  int a = -1001, b = a, c = b;
  int x = 1001, y = x;
  for (int i = 0; i < numsSize; i++) {
    int aa = a, bb = b, xx = x, curr = nums[i];
    a = fmax(a, curr);
    b = fmax(b, fmin(aa, curr));
    c = fmax(c, fmin(bb, curr));
    x = fmin(x, curr);
    y = fmin(y, fmax(xx, curr));
  }
  return fmax(a * b * c, a * x * y);
}

int main() {
  int n1[] = {1, 2, 3};
  int n2[] = {1, 2, 3, 4};
  int n3[] = {-1, -2, -3};
  int r1 = maximumProduct(n1, ARRAY_SIZE(n1));
  int r2 = maximumProduct(n2, ARRAY_SIZE(n2));
  int r3 = maximumProduct(n3, ARRAY_SIZE(n3));
  printf("%d\n", r1);
  assert(r1 == 6);
  printf("%d\n", r2);
  assert(r2 == 24);
  printf("%d\n", r3);
  assert(r3 == -6);
}

File: max_product_3_numbers.py
# 628. Maximum Product of Three Numbers

"""
given an integer array 'nums', find three numbers whose product is maximum
and return the maximum product
"""


class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        n = len(nums)
        for i in range(2):
            k = i
            for j in range(i + 1, n):
                if nums[j] < nums[k]:
                    k = j
            nums[i], nums[k] = nums[k], nums[i]
        for i in range(n - 1, max(-1, n - 4), -1):
            k = i
            for j in range(i):
                if nums[j] > nums[k]:
                    k = j
            nums[i], nums[k] = nums[k], nums[i]
        return max(nums[-1] * nums[-2] * nums[-3], nums[-1] * nums[0] * nums[1])


if __name__ == "__main__":
    obj = Solution()
    print(obj.maximumProduct(nums=[1, 2, 3]))
    print(obj.maximumProduct(nums=[1, 2, 3, 4]))
    print(obj.maximumProduct(nums=[-1, -2, -3]))
